Added Parameter Expression functions for Template Optimization pass

nbronn opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR adds the ability to pass circuits with unbound `Parameter`s into the template optimization transpilation pass. At this time, the `.inverse()` method used in each gate class is not fully implemented to handle the necessary inversion that may be required in certain templates.

### Details and comments

Since `Parameter`s may be used in both templates and circuits, these need to be treated separately so that the template `Parameter`s can be replaced by those from the circuit. These are solved for in `template_substitution.py` using `Sympy`, and hence `Parameter`s defined by LaTeX expressions are currently unsupported due to parsing issues. 

Currently, only the inverse of the `RXGate` is supported due to the type-checking cases for `ParameterExpression`s consisting of `Parameter`s, `ParameterExpression`s consisting of numbers, and numbers. This issue is probably better handled in a general context instead of doing this type-checking in each gate definition. 

Similarly, this PR will not pass the `test_unbound_parameters` unit test in `test_template_matching.py` because the inverse gate is defined as a `UnitaryGate` that is currently cast to a complex `numpy` array in `unitary.py`.

kdk commented on qiskit/transpiler/passes/scheduling/calibration_creators.py:
```diff
@@ -174,7 +175,11 @@ def get_calibration(self, params: List, qubits: List) -> Schedule:
         """
-        theta = params[0]
+        if isinstance(params[0], ParameterExpression):
+            theta = float(params[0]._symbol_expr)
```
`float(params[0])` should accomplish the same (and will raise a `TypeError` if the `ParameterExpression` has unbound parameters).

kdk commented on test/python/transpiler/test_template_matching.py:
```diff
@@ -316,6 +316,10 @@ def count_cx(qc):
+    @unittest.skip(
+        "Skipping because the inverse gate is defined as a UnitaryGate that is \
+        currently cast to a complex numpy array"
+    )
```
Is there an error generated in this case? Can this be avoided?

jakelishman replied:
I'm not really comfortable merging a PR that breaks tests like this.  I don't think this issue is resolved unless the fix is part of it.

kdk reviewed (commented):
Few questions from a quick look. I'm not an expert on the template matching code, but there were a few things that caught my eye. Can you add a test case?

nbronn reviewed (commented):
Thanks for taking the time to review @lcapelluto! I think I have addressed all your comments at this point.

jakelishman commented on qiskit/circuit/library/standard_gates/rx.py:
```diff
@@ -95,7 +95,16 @@ def inverse(self):
         :math:`RX(\lambda)^{\dagger} = RX(-\lambda)`
         """
-        return RXGate(-self.params[0])
+        if isinstance(self.params[0], ParameterExpression):
+            if self.params[0].parameters:
+                param_dict = {}
+                for param in self.params[0].parameters:
+                    param_dict[param] = param.name
+                return RXGate(ParameterExpression(param_dict, -self.params[0].get_sympy_expr()))
+            else:
+                return RXGate(-float(self.params[0].get_sympy_expr()))
+        else:
+            return RXGate(-self.params[0])
```
I still don't understand why `RXGate(-self.params[0])` isn't identical here?  The additional code doesn't appear to do anything other than just negate the `ParameterExpression` in a very verbose manner, and it's not at all clear to me why this should occur in `RXGate.inverse`, and not `ParameterExpression.__neg__`.

If the issue if that there may be parameters that were already bound, then it should `try: -float(self.params[0]); except TypeError` (no need to `get_sympy_expr()`), and it really shouldn't be accessing private members of `ParameterExpression`, but really this code should not be in `RXGate` at all - that would be a bug we could consider fixing in `ParameterExpression`.

jakelishman commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -526,6 +526,12 @@ def is_real(self):
                 return False
         return True
 
+    def get_sympy_expr(self):
+        """Return symbolic expression from sympy"""
+        from sympy import sympify
+
+        return sympify(self._symbol_expr)
+
```
We use `symengine` preferentially here, because `sympy` can be very slow.  We probably don't want to throw that away.

nbronn replied:
Included code to use `symengine` if available.

jakelishman commented on qiskit/dagcircuit/dagcircuit.py:
```diff
@@ -216,7 +216,17 @@ def add_calibration(self, gate, qubits, schedule, params=None):
         if isinstance(gate, Gate):
-            self._calibrations[gate.name][(tuple(qubits), tuple(gate.params))] = schedule
+            try:
+                gate_has_param_exp = isinstance(gate.params[0], ParameterExpression)
```
There's no guarantee that because the 0th parameter is _not_ a `ParameterExpression` that another parameter won't be.

nbronn replied:
Simplified the code so it's not dependent on `gate.params`.

jakelishman commented on qiskit/dagcircuit/dagcircuit.py:
```diff
@@ -216,7 +216,17 @@ def add_calibration(self, gate, qubits, schedule, params=None):
+            except IndexError:
+                gate_has_param_exp = False
+
+            if gate_has_param_exp:
+                self._calibrations[gate.name][
+                    (tuple(qubits), tuple([float(gate.params[0]._symbol_expr)]))
+                ] = schedule
+            else:
+                self._calibrations[gate.name][(tuple(qubits), tuple(gate.params))] = schedule
```
This looks very targetted at only gates with a single parameter.  We need to be general if we're including something like this.  Also, there's no need to access private members of the class - `float(gate.params[0])` is sufficient.

nbronn replied:
Simplified this code so it's not dependent on number of parameters.

jakelishman reviewed (dismissed):
Sorry for slowness here.  I'm not at all an expert in the template optimisation code here, but I'm not really comfortable with how the `RXGate.inverse` seems to be doing a _lot_ of work that is really a very long way out of its concern.  Quite a bit of the templating code also seems to strongly assume that the only gates that should be considered are gates of one parameter, while we have multi-parameter gates.

I think some of the cases where we're going deep into the private methods of classes here are attempts to work around a bug which should be fixed in #6802, and most of the rest are trying to work around bug #6812.  I'm not really comfortable merging new-feature code that introduces very heavy workarounds in lots of unexpected places; I think we need to fix the bugs, and then we can consider doing something like this.

*edit*: by the way, you don't need to press the "update branch" button unless there are merge conflicts indicated.  When the PR is actually marked for merge, the bot will just update it once, which eases our CI usage.  (It's not a _huge_ problem to press the "update" button, it just can tie up our CI a little sometimes.)

nbronn commented:
We added checks for unitarity that now work, but depend on whether `symengine` is available or not (in order to provide a value for `HAS_SYMENGINE`, `symengine` is explicitly used by not needed to avoid tox/lint issues from the `try/except` clause in the preamble). This is a known issue with `symengine` treating real `Symbol`s (https://github.com/symengine/symengine.py/issues/286), so our solution avoids specifying a real `Symbol` if `symengine` imports, and tests for unitarity by binding an arbitrary `float` to unbound parameters. If `symengine` is not available, we use `sympy` to verify symbolically that the gate is unitary (where any `Symbol`s explicitly created are specified with `real=True`). 

Additionally a new unit test called `test_unbound_parameters_in_rzx_template` is added to `test.python.transpiler.test_template_matching` that confirms template matching happens correctly when unbound `Parameter`s are passed through. This was found to be an issue when `sympy.solve()` could not find solutions depending on whether `symengine` was included or not, as for the reasons above.

jakelishman commented:
Sorry, I haven't had time to look through properly, but I'm really concerned about setting `Parameter` to be real-only (symengine is the default, not the special case) - I think that's a really large change, and certainly one we can't make without a very serious deprecation period.  I would need to check to be 100% sure, but I think there are various places in the code (`pulse` and `opflow` come to mind as candidates) that use complex parameters as well, so it might not even be a change we could make if we wanted to.  For example, #7265 fairly recently has code that mentions complex parameters.

I'll have to have more of a look later, but I'm worried that in its current form, this PR is going to have large repercussions across other `Parameter` usage in Terra.

nbronn commented:
> Sorry, I haven't had time to look through properly, but I'm really concerned about setting `Parameter` to be real-only (symengine is the default, not the special case) - I think that's a really large change, and certainly one we can't make without a very serious deprecation period. I would need to check to be 100% sure, but I think there are various places in the code (`pulse` and `opflow` come to mind as candidates) that use complex parameters as well, so it might not even be a change we could make if we wanted to. For example, #7265 fairly recently has code that mentions complex parameters.
> 
> I'll have to have more of a look later, but I'm worried that in its current form, this PR is going to have large repercussions across other `Parameter` usage in Terra.

We reverted the changes that made `Parameter`s real and simply changed the check for unitarity to test bind a real number so that the `test_unbound_parameters` unit test would pass. This should allow flexibility in `Parameter` types while maintaining the check that template optimization continues to work with `Unitary` gates with unbound `Parameter`s.

jakelishman commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -500,39 +502,51 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
-                        bound_params.append(float(param.bind(fake_bind)))
-                    except KeyError:
-                        return None
+                    for key in fake_bind:
+                        bound_params.append(param.assign(key, fake_bind[key]))
```
`for key, value in fake_bind.items():`, but also, isn't this going to add a whole lot of partially bound parameters to the `bound_params` list?  If `fake_bind` has more than one item in it, this is going to be called a whole lot of times, and there'll be more items in `bound_params` than there originally were in `node.op.params`.

nbronn replied:
I think the issue here is that we cannot pass `bound_params` to `.bind()` because of the requirement the bound values are numeric, whereas we cannot pass `bound_params` directly to `.assign()` because it takes `parameter` and `value` separately as arguments. Any obvious solution seems to require modification of how `.bind()` or `.assign()` work.

jakelishman replied:
I don't think this comment matches up with my issue (also the point about iterating through a dictionary with `.items()` still stands).  I think this code is buggy, and it's just luckily not triggered.

The problem is that if `fake_bind` has `n` elements in it, the length of `bound_params` would be `n * len(params)`, which is incorrect.  Logically, this code is trying take `param`, and bind all instances of `Parameter` within it to a corresponding value in `fake_bind`, and return a _single_ value, but instead it adds lots of values to `bound_params`.  This only works because the templates you're considering only have a single `Parameter` in them.  If we had a template that had two parameters in it, this would be broken.  It's quite possible it was broken before as well, but this breaks it in a new way.

Please can you add a test case that involves a template (arbitrary - it can just be invented for the test) that has two parameters, and test that it binds correctly to input circuits with either numeric values or parameter values?  For example, here's a (nonsensical) template match that fails with a `KeyError`, even though it should bind just fine:
```python
In [5]: from qiskit.circuit import Parameter, QuantumCircuit
   ...: from qiskit.transpiler import PassManager
   ...: from qiskit.transpiler.passes import TemplateOptimization, RZXCalibrationBuilder
   ...:
   ...: a = Parameter('a')
   ...: b = Parameter('b')
   ...:
   ...: qc = QuantumCircuit(2)
   ...: qc.p(0.1, 0)
   ...: qc.p(-0.2, 1)
   ...:
   ...: template = QuantumCircuit(2)
   ...: template.p(a, 0)
   ...: template.p(b, 1)
   ...: template.cx(0, 1)
   ...:
   ...: pass_ = TemplateOptimization(
   ...:     user_cost_dict={"cx": 0, "p": 0},
   ...:     template_list=[template],
   ...: )
   ...: PassManager(pass_).run(qc).draw()
```

There are two issues at play: one is the issue with the length of `bound_params`, which is new in this PR, and the other is that the calls to `ParameterExpression.assign` during the template match should assign all parameters in that expression in one go, and _only_ those parameters (this issue I think was pre-existing).

jakelishman replied:
If it _really_ isn't possible to have it work for multiple parameters in the template circuit, then this should at least _enforce_ that the template circuit only has a single unbound parameter in it, and the pass should loudly fail if it's given a template with more than one parameter.

jakelishman commented on releasenotes/notes/add-parameters-to-template-substitution-a1379cdbfcc10b5c.yaml:
```diff
@@ -0,0 +1,88 @@
+---
+features:
+  - |
+    The Parameter class is now allowed in template optimization if the
+    parameterized gates are RZXGates and RXGates. It has been removed from 
```
By `Parameter` we probably mean `ParameterExpression`, I think - that's the more general case.  These can also be Sphinx cross-referenced: `` :class:`.ParameterExpression` ``, ``:class:`.RZXGate`\ s``, etc.

nbronn replied:
Added the cross-referencing

jakelishman commented on releasenotes/notes/add-parameters-to-template-substitution-a1379cdbfcc10b5c.yaml:
```diff
@@ -0,0 +1,88 @@
+  - |
+    The Parameter class is now allowed in template optimization if the
+    parameterized gates are RZXGates and RXGates. It has been removed from 
+    the _remove_impossible() method for scenarios that are incompatible 
```
We shouldn't talk about private implementations in the public release notes.

nbronn replied:
Removed

jakelishman commented on releasenotes/notes/add-parameters-to-template-substitution-a1379cdbfcc10b5c.yaml:
```diff
@@ -0,0 +1,88 @@
+      cal_qc = PassManager(pass_).run(qc_cz.bind_parameters({phi: 0.12}))
+      
+      # Transpile to the backend basis gates
+      cal_qct = transpile(cal_qc, backend)
+      qct = transpile(qc.bind_parameters({phi: 0.12}), backend)
+      
+      # Compare the schedule durations
+      print('Duration of schedule with the calibration:')
+      print(schedule(cal_qct, backend).duration)
+      print('Duration of standard with two CNOT gates:')
+      print(schedule(qct, backend).duration)
+          
+    outputs
+
+    .. parsed-literal::
+
+      Original circuit:
+      			 
+      q_0: ──■──────────────■──
+           ┌─┴─┐┌────────┐┌─┴─┐
+      q_1: ┤ X ├┤ P(2*φ) ├┤ X ├
+           └───┘└────────┘└───┘
+      ZX based circuit:
+      					 ┌─────────────┐            »
+      q_0: ────────────────────────────────────┤0            ├────────────»
+           ┌──────────┐┌──────────┐┌──────────┐│  Rzx(2.0*φ) │┌──────────┐»
+      q_1: ┤ Rz(-π/2) ├┤ Rx(-π/2) ├┤ Rz(-π/2) ├┤1            ├┤ Rx(-2*φ) ├»
+           └──────────┘└──────────┘└──────────┘└─────────────┘└──────────┘»
+      «                                                     
+      «q_0: ────────────────────────────────────────────────
+      «     ┌──────────┐┌──────────┐┌──────────┐┌──────────┐
+      «q_1: ┤ Rz(-π/2) ├┤ Rx(-π/2) ├┤ Rz(-π/2) ├┤ P(2.0*φ) ├
+      «     └──────────┘└──────────┘└──────────┘└──────────┘
+      Duration of schedule with the calibration:
+      1600
+      Duration of standard with two CNOT gates:
+      6848
+      
+    Note that the Parameters/ParameterExpressions must be bound before the 
+    RZXCalibrationBuilder can define calibrations for the RZXGates.
```
The release notes for different sections aren't grouped, so referring to "this PR" will lose the context.  Examples for the PR can go with the feature note.

We don't need to explain where the example came from, just the example is enough.  It'd be good to remove the dependence on `IBMQ` and `load_account` so we're not executing network code on users' hardware if they choose to run this; it'd be better to use a mock backend (e.g. `from qiskit.test.mock import FakeCasablanca`).

nbronn replied:
Fixed the release notes to just give the example, include only a mock backend, eliminate the issues that have since been resolved, and update some transpiler imports.

jakelishman commented on releasenotes/notes/add-parameters-to-template-substitution-a1379cdbfcc10b5c.yaml:
```diff
@@ -0,0 +1,88 @@
+      print('ZX based circuit:')
+      print(qc_cz)
+      
+      # Add the calibratiocns
```
typo

jakelishman commented on test/python/transpiler/test_template_matching.py:
```diff
@@ -353,9 +354,24 @@ def template():
         pass_ = TemplateOptimization(template_list=[template()])
         circuit_out = PassManager(pass_).run(circuit_in)
 
-        # This template will not fully match as long as gates with parameters do not
-        # commute with any other gates in the DAG dependency.
-        self.assertEqual(circuit_out.count_ops().get("cx", 0), 2)
+        self.assertEqual(circuit_out.count_ops().get("cx", 0), 0)
```
I know this was pre-existing, but it'd be good to make this a test of the entire structure of the circuit, not just an op-count - as I understand it, this test should now return an empty circuit, so we should test that.

(As an aside: it maybe would be good to update the test slightly so it doesn't return a fully empty circuit, because a buggy feature completely wiping out all data in the circuit seems like a plausible failure mode.)

nbronn replied:
I was attempting to improve this test by using `numpy.testing.assert_allclose` and passing it the `circuit_out` with bound parameters through the `Operator` class from `qiskit.quantum_info`, and ran right into an [issue I raised](https://github.com/Qiskit/qiskit-terra/issues/6633) dealing with `Parameter`s not being removed from the parameter table. Since this is not yet fixed, I do not think there is a good way to solve this currently.

Perhaps we could move this point to another Issue?

jakelishman replied:
I looked more, and this test is vitally important, and this PR breaks it as it is.  The output of the test _should_ contain 2 CX gates.  If you run this test with the changes in this PR, you get
```python
In [4]: circuit_out.draw()
Out[4]:
     ┌────┐
q_0: ┤0   ├─────────────
     │  P │┌───────────┐
q_1: ┤1   ├┤ P(-1.0*β) ├
     └────┘└───────────┘
```
Note that it's replaced the two CXs with two gates whose parameters don't mean anything.  This is a failure; the pass should not have replaced anything, even with your changes.  If the test had stronger assertions (like I was suggesting in the previous comment), it would have been clearer that this was a true test failure.

Fortunately, now that I understand this, we can hugely simplify the requirements on `UnitaryGate`; we can rewrite the test to avoid it by just using a gate that can handle `Parameter` in its inverse.  Here's a clearer version of the test that's exactly equivalent:

```python
    def test_optimizer_does_not_replace_unbound_partial_match(self):
        beta = Parameter("β")
        template = QuantumCircuit(2)
        template.cx(1, 0)
        template.cx(1, 0)
        template.p(beta, 1)
        template.cu(0, 0, 0, -beta, 0, 1)

        circuit_in = QuantumCircuit(2)
        circuit_in.cx(1, 0)
        circuit_in.cx(1, 0)
        pass_ = TemplateOptimization(
            template_list=[template],
            user_cost_dict={"cx": 6, "p": 0, "cu": 8},
        )

        circuit_out = PassManager(pass_).run(circuit_in)

        # The template optimisation should not have replaced anything, because
        # that would require it to leave dummy parameters in place without
        # binding them.
        self.assertEqual(circuit_in, circuit_out)
```

If we change the test to that, it lets us revert all the changes to `UnitaryGate` that I now see are unrelated (sorry again about that), and the test is both clearer and stricter.

The trick, though, is that I think to achieve your goals in the pass, you're going to need to update your logic a bit.  You need to keep track of which parameters are dummy parameters used in the template, and which are parameters that come from the circuit.  For example, in this test, the `beta` should not be allowed to be output by the transpiler pass, but in the test you added after this one, the `phi` _should_ be output (as you have it).

There's more nuance than just that as well, though, because the pass needs to handle the case that a template uses a parameter with the same name as an input circuit without breaking.  You can likely handle that bit by re-binding the variables in a template to auto-generated unique names if you detect issues.

nbronn replied:
That's a great point to eliminate the `UnitaryGate` from the unit test entirely (I implemented the above, and also modified `test_parametric_template` similarly). The `test_unbound_parameters_in_rzx_template` test is now working so that `circuit_in` and `circuit_out` are verified to not be equal (because the template optimization has changed the circuit), yet the operators are equivalent after binding `phi` to a `float`.

In order to guarantee template optimization does *not increase* the number of `Parameter`s in a circuit, there is a new private method `_incr_num_parameters` in `template_substitution.py` that counts the number of `Parameter`s in the `template` and in the `circuit` and returns true if the former is greater than the latter. If this is true, logic inside `_substitution` will continue out of the method, preventing the template from being substituted.

As for `Parameter`s in circuits and templates that have the same name, I'm not sure I know how to auto-generate unique names. Trying this out on the test notebook, all the `template`s I use are of the `rzx` variety, and the specific `Parameter` used is always
```
theta = Parameter("ϴ") # used in all rzx templates
```
which does not match with
```
theta = Parameter('θ') # typed from my "Greek" keyboard
```
or
```
theta = Parameter('$\\theta$') # maybe the safest, LaTeX version
```
so it seems that this is not much of a problem currently. (Of course template optimization fails if I use the exact form of `theta` from the template).

jakelishman commented on test/python/transpiler/test_template_matching.py:
```diff
@@ -353,9 +354,24 @@ def template():
-        self.assertEqual(circuit_out.count_ops().get("cx", 0), 2)
+        self.assertEqual(circuit_out.count_ops().get("cx", 0), 0)
+
+    def test_unbound_parameters_in_rzx_template(self):
+        """
+        Test that rzx template ('zz2') functions correctly for a simple
+        circuit with an unbound ParameterExpression.
+        """
+
+        phi = Parameter("$\\phi$")
+        circuit_in = QuantumCircuit(2)
+        circuit_in.cx(0, 1)
+        circuit_in.p(2 * phi, 1)
+        circuit_in.cx(0, 1)
+
+        pass_ = TemplateOptimization(**rzx_templates(["zz2"]))
+        circuit_out = PassManager(pass_).run(circuit_in)
+
+        self.assertEqual(circuit_out.count_ops().get("cx", 0), 0)
```
Is there an exact expected output of this circuit?  It'd be good to really make the assertions as tight as possible, to test that the optimization has completely succeeded - at present, if it just returned an empty circuit it'd still pass this test.

Also, can we add some more complex tests when parameters are parameter expressions with multiple different symbols in them as well?

nbronn replied:
Fixed

jakelishman replied:
This isn't actually fixed - it's got the same assertions as before.  A good test would be to construct the exact output circuit you expect, and then do `self.assertEqual(circuit_out, expected)`.

nbronn replied:
Fixed this time!

jakelishman reviewed (commented):
I'm sorry that I took so long to look at this - I've come and gone on it a little bit, especially because I'm very nervous when reviewing code around `ParameterExpression`.

This looks to me like it's in a much better form than the original implementation, and I'm moving towards being happy to accept it.  I have some suggestions inline - most are rather minor, but I think I identified some buggy behaviour which it'd be great to expand the tests to cover, and there's a couple of places where I think we can simplify things quite a bit.

nbronn commented on qiskit/extensions/unitary.py:
```diff
@@ -56,11 +75,28 @@ def __init__(self, data, label=None):
+                for jj in range(numpy.shape(data)[1]):
+                    if isinstance(data[ii][jj], ParameterExpression):
+                        for param in data[ii][jj].parameters:
+                            data[ii][jj] = data[ii][jj].bind({param: float(0.42)})
```
Completely makes sense, added a print statement about determining the type of unbound `Parameter`s and `pass`.

rafal-pracht commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -532,6 +538,15 @@ def is_real(self):
 
+    def get_sympy_expr(self):
+        """Return symbolic expression from sympy/symengine"""
+        if HAS_SYMENGINE:
+            return symengine.sympify(self._symbol_expr)
+        else:
+            from sympy import sympify
+
+            return sympify(self._symbol_expr)
```
Renamed to 'to_simplify_expression', this function returns the simplified version of the expression.

rafal-pracht commented on qiskit/transpiler/passes/calibration/builders.py:
```diff
@@ -224,10 +224,17 @@ def get_calibration(self, node_op: CircuitInst, qubits: List) -> Union[Schedule,
+        try:
+            theta = float(node_op.params[0])
+        except TypeError as ex:
+            raise QiskitError(
```
Fixed

rafal-pracht commented on qiskit/transpiler/passes/calibration/builders.py:
```diff
@@ -224,10 +224,17 @@ def get_calibration(self, node_op: CircuitInst, qubits: List) -> Union[Schedule,
+                the qubits.
         """
-        theta = node_op.params[0]
+        try:
+            theta = float(node_op.params[0])
+        except TypeError as ex:
+            raise QiskitError(
+                "This transpilation pass requires all Parameters to be bound."
+            ) from ex
```
Fixed

rafal-pracht commented on qiskit/extensions/unitary.py:
```diff
@@ -30,10 +30,29 @@
 _DECOMPOSER1Q = OneQubitEulerDecomposer("U3")
 
 
+def to_sympy_matrix(numpy_matrix):
+    """Convert (numpy) matrix to sympy matrix."""
+    matrix = []
+    i = 0
+    for row in numpy_matrix:
+        nrow = []
+        i += 1
+        for expr in row:
+            if isinstance(expr, ParameterExpression):
+                expr = expr.get_sympy_expr()
+            nrow.append(expr)
+        matrix.append(nrow)
+
+    from sympy import Matrix
+
+    return Matrix(matrix), i
```
The method is removed, now we don't check unitary when the parameter is unbound.

rafal-pracht commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -500,39 +502,51 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
-
-        if not symbols:
+        equations, circ_dict, temp_symbols, sol, fake_bind = [], {}, {}, {}, {}
+        for t_idx, template_params in enumerate(template_params):
```
Fixed

rafal-pracht commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -500,39 +502,51 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
+                if isinstance(circuit_params[t_idx], ParameterExpression):
+                    circ_param_sym = circuit_params[t_idx].get_sympy_expr()
+                else:
+                    circ_param_sym = parse_expr(str(circuit_params[t_idx]))
```
Not necessary, this can be an unbound parameter.

rafal-pracht commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -500,39 +502,51 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
+                    circ_param_sym = circuit_params[t_idx].get_sympy_expr()
+                else:
+                    circ_param_sym = parse_expr(str(circuit_params[t_idx]))
+                equations.append(sym.Eq(template_params.get_sympy_expr(), circ_param_sym))
```
The name of the method was incorect, when I've created it I didn't know about symengine and I assumed that all symbols come from simpy. Now it is renamed to 'to_simplify_expression',

rafal-pracht commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -500,39 +502,51 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
+                    circ_param_sym = parse_expr(str(circuit_params[t_idx]))
+                equations.append(sym.Eq(template_params.get_sympy_expr(), circ_param_sym))
+
+                for param in template_params.parameters:
+                    if not HAS_SYMENGINE:
+                        temp_symbols[param] = sym.Symbol(str(param))
+                    else:
+                        temp_symbols[param] = symengine.Symbol(str(param))
+
+                if isinstance(circuit_params[t_idx], ParameterExpression):
+                    for param in circuit_params[t_idx].parameters:
+                        if not HAS_SYMENGINE:
+                            circ_dict[param] = sym.Symbol(str(param))
+                        else:
+                            circ_dict[param] = symengine.Symbol(str(param))
```
Fixed

rafal-pracht commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -500,39 +502,51 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
 
         # Check compatibility by solving the resulting equation
-        sym_sol = sym.solve(equations)
+        sym_sol = sym.solve(equations, set(temp_symbols.values()))
```
without set, this crash

rafal-pracht commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -35,6 +35,12 @@
 ParameterValueType = Union["ParameterExpression", float]
 
 
+class ParameterTypeError(TypeError):
+    """The type error throw by parameter if we try to cast to number unbind parameter."""
+
+    pass
```
I've move this to qiskit.circuit.exceptions, but I'm unable to inherit from QiskitError because in that case, the code doesn't work. Let me know if this is needed. In that case, I will try to understand why this happens.

jakelishman commented on qiskit/extensions/unitary.py:
```diff
@@ -56,11 +58,19 @@ def __init__(self, data, label=None):
-        # Convert to numpy array in case not already an array
-        data = numpy.array(data, dtype=complex)
-        # Check input is unitary
-        if not is_unitary_matrix(data):
-            raise ExtensionError("Input matrix is not unitary.")
+
+        # Check if there is any unbound parameter in array
+        try:
+            # Convert to numpy array in case not already an array
+            data = numpy.array(data, dtype=complex)
+
+            # Check input is unitary
+            if not is_unitary_matrix(data):
+                raise ExtensionError("Input matrix is not unitary.")
+        except ParameterTypeError:
+            print("Type of unbound Parameter cannot be determined to ensure unitarity.")
+            pass
+
```
I'm sorry for my not understanding before, and for the mess.  I think we should revert all the changes to `UnitaryGate`; I understand the test better now, and I recognise that `UnitaryGate` isn't the problem at all, it's that the input test is rather strange.

nbronn replied:
Reverted changes.

jakelishman reviewed (commented):
I'm sorry that I've led you on a bit of a goose chase with respect to `UnitaryGate` - I didn't fully understand what the pass was doing at first, nor what your changes were trying to do.  I understand now, and the changes to `UnitaryGate` are something we should drop from this PR (sorry).

More importantly, though, the test that's modified by this PR is indicative of a logic problem with this PR.  I've [explained more further down](https://github.com/Qiskit/qiskit-terra/pull/6899#discussion_r798020248), but there's more logic that needs to be tracked through the pass.  The issues with `UnitaryGate` aside (I've written a new version of the test for you that no longer needs it), that test needs to pass without modifying its assertions - it _should_ have two cx gates in it afterwards (but my new assertion is better).

Now that I understand that, though, I'm more confident in my ability to review this.  Sorry it took me so long (and a couple of review rounds) to get to that point, though.

rafal-pracht commented on qiskit/circuit/exceptions.py:
```diff
@@ -19,3 +19,9 @@ class CircuitError(QiskitError):
     pass
+
+
+class ParameterTypeError(TypeError):
+    """The type error throw by parameter if we try to cast to number unbind parameter."""
+
+    pass
```
I didn't express myself clearly. I know that I can have multiple inheritances, and it works correctly in the general case. But not in our case. The issue lay in the operator.py and in catching the QiskitError there. If ParameterTypeError inherits from Qiskit error it is caught there and not in template optimization. And in a result, the tests fail.

![image](https://user-images.githubusercontent.com/55279376/152678821-6fa0cf58-ddf9-44fb-ad10-0f6afb0c66d2.png)

rafal-pracht replied:
So If we would like to inherit from QiskitError I have to change the operator.py and catch fewer general exceptions or change in some way our code in template optimization. Therefore I prefer to inherit only from type errors. @jakelishman  Do you agree?

jakelishman replied:
Ah, I understand your issue - that's pretty awkward programming in `operator.py` to be honest, and I wouldn't really know what to suggest.  Fortunately, I think we can sidestep a lot of this, because the changes to `UnitaryGate` became unnecessary, so we can just remove the exception as well.

jakelishman reviewed (commented):
This is definitely getting closer.  Please could you really try and write some more stringent tests; really try and think of the edge cases of the template optimisation pass, and write tests that try as hard as they can to reach the limits of those cases.  I've suggested several both in the comments in this review, and in prior review comments - this PR adds a big new feature to the pass, and we really need to heavily test it.  It's not just enough to test the couple of paths corresponding to your expected uses, we have to make sure it works right beyond that too, and equally importantly that it _fails_ loudly and correctly when the input is invalid in some form or another.

nbronn commented on releasenotes/notes/add-parameters-to-template-substitution-a1379cdbfcc10b5c.yaml:
```diff
@@ -0,0 +1,78 @@
+      from qiskit.transpiler import PassManager
+      from qiskit.transpiler.passes import TemplateOptimization
+      
+      # New contributions to the template optimization
+      from qiskit.transpiler.passes.scheduling import RZXCalibrationBuilder, rzx_templates
```
Fixed that in the release notes.

nbronn commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -550,6 +535,15 @@ def is_real(self):
                 return False
         return True
 
+    def to_simplify_expression(self):
+        """Return symbolic expression from sympy/symengine"""
+        if _optionals.HAS_SYMENGINE:
+            return symengine.sympify(self._symbol_expr)
+        else:
+            from sympy import sympify
+
+            return sympify(self._symbol_expr)
+
```
Fixed to not explicitly use `sympify` and changed name of method to `to_sympify_expression()`.

nbronn commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -450,7 +436,7 @@ def __complex__(self):
         # TypeError is for sympy, RuntimeError for symengine
         except (TypeError, RuntimeError) as exc:
-            raise TypeError(
+            raise ParameterTypeError(
```
Reverted error type back to `TypeError` and removed `ParameterTypeError`.

nbronn commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -19,9 +19,15 @@
+from qiskit.circuit.exceptions import CircuitError, ParameterTypeError
 from qiskit.utils import optionals as _optionals
 
+try:
+    import symengine
+
+    _optionals.HAS_SYMENGINE = True
+except ImportError:
+    _optionals.HAS_SYMENGINE = False
```
Modified check to import `symengine` if `_optionals.HAS_SYMENGINE` is `True`.

nbronn commented on test/python/transpiler/test_template_matching.py:
```diff
@@ -308,54 +310,69 @@ def count_cx(qc):
-        circuit_in.cx(1, 0)
-        circuit_in.cx(1, 0)
+        circuit_in.cx(0, 1)
+        circuit_in.p(2 * phi, 1)
+        circuit_in.cx(0, 1)
 
-        pass_ = TemplateOptimization(template_list=[template()])
+        pass_ = TemplateOptimization(**rzx_templates(["zz2"]))
         circuit_out = PassManager(pass_).run(circuit_in)
 
-        # This template will not fully match as long as gates with parameters do not
-        # commute with any other gates in the DAG dependency.
-        self.assertEqual(circuit_out.count_ops().get("cx", 0), 2)
+        # these are NOT equal if template optimization works
+        self.assertNotEqual(circuit_in, circuit_out)
```
I started coupling many of the template parameter tests with an `assertNotEqual()` to verify that the circuit was modified from it's original form and `assertEqual()` that binds the `Parameter`s to arbitrary floats and uses `Operator` to make sure they are the same unitary (up to a global phase).

nbronn commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -508,42 +501,67 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
 
         return template_dag_dep
+
+    def _incr_num_parameters(self, template):
+        """
+        Checks if template substitution would increase the number of
+        parameters in the circuit.
+        """
+        template_params = set()
+        for param_list in (node.op.params for node in template.get_nodes()):
+            for param_exp in param_list:
+                if isinstance(param_exp, ParameterExpression):
+                    template_params.update(param_exp.parameters)
+
+        circuit_params = set()
+        for param_list in (node.op.params for node in self.circuit_dag_dep.get_nodes()):
+            for param_exp in param_list:
+                if isinstance(param_exp, ParameterExpression):
+                    circuit_params.update(param_exp.parameters)
+
+        return len(template_params) > len(circuit_params)
```
This is now resolved in `_attempt_bind()`, where if a template `Parameter` matches that of a circuit, a new `Parameter` with a variable name is created and the nodes in the `template_dag_dep` are overwritten with these new "substituted" `Parameter`s. The test `test_unbound_parameters_in_rzx_template` now also checks that this behavior works.

nbronn commented:
Concerning the main remaining [issue](https://github.com/Qiskit/qiskit-terra/pull/6899#discussion_r811378227) here, I fixed the logic so that the binding attempts go through every (assumed) `ParameterExpression` in `node.op.params`, then every `Parameter` in each of those `ParameterExpression`s is bound if it is in `fake_bind` (i.e., has a solution from the `sympy` solver). This should solve the problem with the length of `bound_params` getting out of hand.

It also works with multiple parameters, and I've added a test `test_two_parameter_template` to `test_template_matching.py` that successfully achieves two-parameter template matching.

nbronn commented on qiskit/circuit/library/probability_distributions/__init__.py:
```diff
@@ -0,0 +1,11 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
```
Deleted!

jakelishman commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -492,30 +502,76 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
+                                random.choice(string.ascii_lowercase) for i in range(8)
+                            )
+                            sub_params[t_param] = Parameter(new_param_name)
+                            t_param_exp = t_param_exp.assign(t_param, sub_params[t_param])
```
This generates a new parameter every time the clash is encountered, which would likely have caused issues if the same parameter needed to be put into a circuit more than once.  Better to create the new parameters only once, and retrieve them from a dictionary on subsequent accesses.  This can be made quite elegant with `collections.defaultdict` - I've done that in a14dc37441f73e8c9630d43b03ba8a1c9de5c757.

jakelishman commented on qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:
```diff
@@ -492,30 +502,76 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
+                            template_params.append(t_param_exp)
+                        else:
+                            sub_node_params.append(t_param_exp)
+                            template_params.append(t_param_exp)
```
These appends were at the wrong level - they only got added to the lists if the parameters were expressions.  That was a problem for `template_params` in particular, since it would have the wrong number of elements, and the subsequent zip between `circuit_params` and `template_params` would line things up incorrectly.

I added very explicit tests to show the failures in fc8315688ffb5ebca7e56c9b6f72887e90879dea, and fixed them in a14dc37441f73e8c9630d43b03ba8a1c9de5c757.

jakelishman reviewed (commented):
I think the recent changes had broken some of the core template matching - in my tests, very simple templates were inserting the wrong numeric parameters in the pass.  This is really why I've been hammering on the idea of _exact_ tests of the behaviour - I know the test suite was lacking in them before, but they really are crucial.

This PR's been going for a while now, in good part because of slowness on my part and me being away, so I've gone ahead and fixed the bugs I found in this review, added the explicit tests I'd been asking for and improved a couple of the data structures used in the matching.  I'll have another look tomorrow before merging to make sure to check my own code, but unless I've made a howler in the test suite, this is hopefully good to merge.

Thanks for sticking with it - it'd be good if you could look over the new tests that I've added to make sure you agree that they all should pass (they currently do). If so, I'll be able to merge it soon.

mtreinish commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -22,6 +22,8 @@
 from qiskit.utils import optionals as _optionals
 
+if _optionals.HAS_SYMENGINE:
+    import symengine
```
Actually symengine isn't that heavy to import as it's mostly cython and a statically linked c++ lib. It's sympy we normally have to avoid as it's super slow to import. The bigger thing is that symengine has a narrower platform support than qiskit and we can't rely on it being present, it's in the requirements list for platforms that support it but not all do.

jakelishman replied:
Oh, it actually appeared quite heavy on my laptop when I tried it - must have hitting a really cold cache.  At any rate, these changes were unrelated to this PR logically, so I put them back in place.

nbronn commented:
Thanks @jakelishman! I can confirm that all the tests in `test.python.transpiler.test_template_matching.TestTemplateMatching` pass on my laptop as well.
